File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number[]}
 */
var nodesBetweenCriticalPoints = function(head) {
    let extremeIndices = [];
    
    let beforeMiddle = head.val;
    let middle = head.next.val;
    let middleIndex = 1;
    
    head = head.next;
    head = head.next;
    
    while(head != null) {
        if(middle > beforeMiddle && middle > head.val) {
            extremeIndices.push(middleIndex);
        }
        
        if(middle < beforeMiddle && middle < head.val) {
            extremeIndices.push(middleIndex);
        }
        
        beforeMiddle = middle;
        middle = head.val;
        head = head.next;
        middleIndex++;
    }
    
    if(extremeIndices.length < 2) {
        return [-1, -1];
    }
    
    let minimumDistance = Number.MAX_VALUE;
    
    for(let index = 1; index < extremeIndices.length; index++) {
        minimumDistance = Math.min(minimumDistance, extremeIndices[index] - extremeIndices[index - 1]);
    }
    
    let maximumDistance = extremeIndices[extremeIndices.length - 1] - extremeIndices[0];
    
    return [minimumDistance, maximumDistance];
};

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        List<Integer> indexList = new ArrayList<>();
        
        ListNode previous = head;
        head=head.next;
        int index = 1;
        
        while(head.next != null) {
            if(previous.val < head.val && head.next.val < head.val ) {
                indexList.add(index);
            } else if(previous.val > head.val && head.next.val > head.val ) {
                indexList.add(index);
            }
            
            previous = head;
            head = head.next;
            index++;
        }
        
        if(indexList.size() < 2) {
            int[] empty = new int[2];
            empty[0] = -1;
            empty[1] = -1;
            return empty;
        }
        
        int smallest = Integer.MAX_VALUE;
        
        for(int i = 1; i < indexList.size(); i++) {
            smallest = Math.min(smallest, indexList.get(i) - indexList.get(i - 1));
        }
        
        int largest = indexList.get(indexList.size() - 1) - indexList.get(0);
        
        int[] result = new int[2];
        
        result[0] = smallest;
        result[1] = largest;
        
        return result;
    }
}